sequencer/src/persistence/fs.rs

```diff
@@ -1454,40 +1454,140 @@ impl MembershipPersistence for Persistence {
+            let filename = format!("{}_{}", block_number, event_index);
+            let file_path = events_dir.join(filename).with_extension("json");
 
+            if file_path.exists() {
+                continue;
+            }
+
+            let file = File::create(&file_path).context("Failed to create event file")?;
+            let writer = BufWriter::new(file);
+
+            serde_json::to_writer_pretty(writer, &event)
+                .context("Failed to write event to file")?;
+        }
+
+        // update the l1 block for which we have processed events
         inner.replace(
-            &file_path,
+            &last_l1_finalized_path,
             |_| Ok(true),
-            |file| {
-                let writer = BufWriter::new(file);
-                serde_json::to_writer_pretty(writer, &(l1_block, events))?;
+            |mut file| {
+                let bytes = l1_block.to_le_bytes();
+
+                file.write_all(&bytes)?;
+                tracing::debug!("updated l1 finalized ={l1_block:?}");
                 Ok(())
             },
         )
     }
 
-    async fn load_events(&self) -> anyhow::Result<Option<(u64, Vec<(EventKey, StakeTableEvent)>)>> {
-        let inner = self.inner.write().await;
-        let dir_path = &inner.stake_table_dir_path();
+    async fn load_events(
+        &self,
+        l1_block: u64,
+    ) -> anyhow::Result<(Option<u64>, Vec<(EventKey, StakeTableEvent)>)> {
+        let inner = self.inner.read().await;
+        let dir_path = inner.stake_table_dir_path();
         let events_dir = dir_path.join("events");
 
-        let file_path = events_dir.with_extension("txt");
+        // check if we have any events in storage
+        // we can do this by checking last l1 finalized block for which we processed events
+        let last_l1_finalized_path = events_dir.join("last_l1_finalized").with_extension("txt");
 
-        if !file_path.exists() {
-            return Ok(None);
+        if !last_l1_finalized_path.exists() || !events_dir.exists() {
+            return Ok((None, Vec::new()));
         }
 
-        let file = File::open(file_path)?;
-        let reader = BufReader::new(file);
-        let (l1_block, events) = serde_json::from_reader(reader)?;
-        Ok(Some((l1_block, events)))
+        let mut events = Vec::new();
+
+        let bytes = fs::read(&last_l1_finalized_path).context("read")?;
+        let mut buf = [0; 8];
+        bytes.as_slice().read_exact(&mut buf[..8])?;
+
+        let last_l1 = u64::from_le_bytes(buf);
+
+        // Determine the L1 block for querying events.
+        // If the last stored L1 block is greater than the requested block, limit the query to the requested block.
+        // Otherwise, query up to the last stored block.
+        let query_l1_block = if last_l1 > l1_block {
+            l1_block
+        } else {
+            last_l1
+        };
+
+        for entry in fs::read_dir(&events_dir).context("events directory")? {
```

Comment:
Okay for example here we list all entries.

Reply:
yeah I don't think we can avoid this if we want separate entry  for each event